### PR TITLE
Draft for the legacy component to return custom js code

### PR DESCRIPTION
**Pull request by rsheer**
With this PR, I suggest to add the ability to the legacy component to return custom signals.

To add a custom signal to a legacy component, call `withCustomSignal($signal_name, $js_code)`. Since there can be multiple custom signals in one legacy component, give each signal an id, which is unique in this component. The second parameter is the JavaScript-Code, which should be executed when the signal is triggered.

`getCustomSignal($signal_name)` returns a signal with the custom JavaScript-code in it. If there is no custom signal with the given name, `null` will be returned.

### CODE REVIEW

File: src/UI/Component/Legacy/Legacy.php [hunk 1]
```diff
@@ -16,4 +18,21 @@ interface Legacy extends \ILIAS\UI\Component\Component
...
+     * @param $js_code
+     * @return Legacy
+     */
+    public function withCustomSignal(string $signal_name, string $js_code) : Legacy;
+
+    /**
+     * Get signal with custom JavaScript code
+     *
+     * @param $signal_name
+     * @return Signal|null
```
Comment (nhaagen) [inline src/UI/Component/Legacy/Legacy.php]: I'd rather throw an Exception than silently return null.

**Review by nhaagen — CHANGES_REQUESTED** [review]
Hey @rsheer ,
thanks for your PR.

Please add the methods from the interface to the actual component; 
you may throw new \ILIAS\UI\NotImplementedException(); instead of actually implementing them.
This way the tests will not fail ;)

Please consider throwing an Exception when looking for a non-existent signal. Either way there is something fishy and the triggers will most likely not work, so complain early.

Cheers,
Nils

**Comment by klees** [comment]
Hi @rsheer!

Thanks for this addition to the UI framework. Thx @nhaagen for looking into this with @rsheer.

I feel that I need to add a little context on why this currently seems to be the best solution to combine javascript in legacy components with the signal handling from the UI framework.

If I want to use a UI component, like the Mode View Control, with some legacy component that should act based on the actions taken by the user on the UI components, this currently is near impossible. Since JS code connecting two components basically needs to know IDs from both components, the only mechanism `withAdditionalOnLoadCode` to bind JS code directly to components won't be enough here. This basically is the problem that the signal mechanism in the UI framework solves in the first place. Additionally, the code that actually connects the two components is rendered together with the component that should react, which in this case would be the Legacy Component. The complete signal implementation seems to be solid, as it is (intentionally) next to impossible to interfere here with custom JS code.

Still, we need some way to connect legacy code to the new UI components if we want to transition to the UI components smoothly. Thus, this proposal here. This should not be considered "the" way to include JS code and this will also vanish, hopefully, sometimes soon.

@rsheer: Please add a "DEPRECATED" to the two methods you are introducing, telling people that they should only be used to connect to existing legacy components, are likely to be removed sometime soon and are to be used completely at everyones own risk.

Best regards!

**Comment by rsheer** [comment]
Hi @nhaagen and @klees ,

Thanks a lot for your reviews! I just made the changes according to your feedback.

I agree with you @klees about the reason why this solution is a good idea, but also should be marked as deprecated. In my opinion, this is the very reason why this component is named *Legacy* and not *HTMLContainer* (or something like that). It is here to use components that are not implemented or suggested yet with the other components of the UI framework together.

In this case of the *Mode View Control* in combination with multiple Legacy components, it might be a good idea to create something like *Tab*-component. But I think this should be a topic for another PR if a component like this is desired.

Thanks again for the reviews and feel free to bring some more feedback if there is something to change. :)

Kind Regards
Raphael

PS: It wasn't intended to mark an existing Github user with my latest commit message ;)

File: src/UI/Component/Legacy/Legacy.php (hunk 1 above)
Comment (rsheer) [reply src/UI/Component/Legacy/Legacy.php]: I added now a throws \InvalidArgumentException. This was the best fitting one I had in mind.

**Review by nhaagen — APPROVED** [review]

**Comment by matthiaskunkel** [comment]
Jour Fixe, 21 OCT 2019 : We highly appreciate this suggestion and accept the PR for trunk. Please feel free to merge to trunk.

**Pull request merged**